### PR TITLE
Data re-ordering and operator pruning

### PR DESCRIPTION
**Pull request by wmcclinton**
This PR is made to help solve the weird cases that come up due to our backchaining approach not being data invariant. We do this by (1) re-ordering the dataset during nsrt_learning (settings.py now has parameter data_orderings_to_search = 1 (default), which represents how many orderings we will search over) to find the pnads set with the minimum complexity and (2) pruning harmless operators by increasing min_perc_data_for_nsrts incrementally until we find the smallest harmless operator set.

### CODE REVIEW

File: predicators/nsrt_learning/nsrt_learning_main.py
Comment (NishanthJKumar) [inline predicators/nsrt_learning/nsrt_learning_main.py:42]: change "smallest" to "least complex"

File: predicators/nsrt_learning/nsrt_learning_main.py
Comment (NishanthJKumar) [inline predicators/nsrt_learning/nsrt_learning_main.py:45]: change "data" to "data order"

File: tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py
Comment (NishanthJKumar) [inline tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py:289]: change this comment to say "data order invariant!"

File: tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py
Comment (NishanthJKumar) [inline tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py:594]: wait - I'm a little confused by this: why do the pnad's have different lengths, but the NSRT's work? Could you augment this comment to explain more about what exactly is going on here?

**Review by NishanthJKumar — COMMENTED** [review]
LGTM after addressing comments; great work! I'll let @tomsilver take the next round of review and provide actual approval.

File: predicators/utils.py
Comment (tomsilver) [inline predicators/utils.py:169]: can we name this something more specific?

File: predicators/utils.py
Comment (tomsilver) [inline predicators/utils.py:179]: this way looks a little weird to me. can we instead use the pattern here? https://stackoverflow.com/questions/23289547/shuffle-two-list-at-once-with-same-order

File: tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py
Comment (tomsilver) [inline tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py:294]: necessary*

File: predicators/nsrt_learning/nsrt_learning_main.py
Comment (tomsilver) [inline predicators/nsrt_learning/nsrt_learning_main.py:44]: approaches*

File: predicators/utils.py
Comment (tomsilver) [reply predicators/utils.py:179]: actually, now that I see where this is being used, maybe we don't really need a utility function for this and we can just move this code inside the loop? the fact that it's hard to come up with a specific name is good evidence that we probably won't be using this utility elsewhere. and since it's short, might as well just put it where it's used.

File: predicators/nsrt_learning/strips_learning/base_strips_learner.py
Comment (tomsilver) [inline predicators/nsrt_learning/strips_learning/base_strips_learner.py:46]: cool! I didn't know that this was how you implemented this specifically. I like it

File: predicators/nsrt_learning/strips_learning/base_strips_learner.py
Comment (tomsilver) [inline predicators/nsrt_learning/strips_learning/base_strips_learner.py:50]: iterate*?

File: predicators/nsrt_learning/strips_learning/base_strips_learner.py
Comment (tomsilver) [inline predicators/nsrt_learning/strips_learning/base_strips_learner.py:53]: it would be wiser to only iterate over thresholds that make a difference, right? and since checking harmlessness is slow, this actually seems like a change worth making. can we collect up all the numbers of data across the NSRTs and then only consider thresholding at those numbers?

File: predicators/nsrt_learning/strips_learning/base_strips_learner.py
Comment (tomsilver) [inline predicators/nsrt_learning/strips_learning/base_strips_learner.py:67]: the logic here is a bit confusing to follow for the cases we had before (where harmless op pruning is not enabled). could we refactor this a bit so that all that happens in this for loop is we determine a good `min_perc_data_for_nsrt` to use, and the for loop only gets run in the case where `enable_harmless_op_pruning` is True; and then outside of the for loop, we use the determined `min_perc_data_for_nsrt` to construct the final PNADs?

**Review by tomsilver — COMMENTED** [review]
@wmcclinton really nice work! here are some suggestions

File: predicators/utils.py
Comment (tomsilver) [reply predicators/utils.py:169]: actually, now that I see where this is being used, maybe we don't really need a utility function for this and we can just move this code inside the loop? the fact that it's hard to come up with a specific name is good evidence that we probably won't be using this utility elsewhere. and since it's short, might as well just put it where it's used.

File: predicators/nsrt_learning/strips_learning/base_strips_learner.py
Comment (tomsilver) [inline predicators/nsrt_learning/strips_learning/base_strips_learner.py:77]: the code below looks quite similar to the code above -- is it possible to remove the "else" here and just have the code once?

File: predicators/third_party/fast_downward_translator/axiom_rules.py
Comment (tomsilver) [inline predicators/third_party/fast_downward_translator/axiom_rules.py:12]: looks like the autoformatting went wild here -- can you revert all these third party files?

**Review by tomsilver — COMMENTED** [review]
looking great! a few small things and then we should be good to merge

**Comment by tomsilver** [comment]
also, let's rerun the sidelining experiments after we merge this

File: predicators/nsrt_learning/strips_learning/base_strips_learner.py
Comment (tomsilver) [inline predicators/nsrt_learning/strips_learning/base_strips_learner.py:70]: is this going to work as expected if you set `--min_data_for_nsrt` to be something nontrivial, but then also `--enabled_harmless_op_pruning`?

Comment (wmcclinton) [reply predicators/nsrt_learning/strips_learning/base_strips_learner.py:70]: It will always chose the max between min_data_for_nsrt and the min_perc_data_for_nsrt. So for all min_perc_data_for_nsrts it is trying below the min_data_for_nsrt it will just use the min_data_for_nsrt instead. If min_data_for_nsrt causes us to break harmlessness it will return the original learned pnads.

Comment (tomsilver) [reply predicators/nsrt_learning/strips_learning/base_strips_learner.py:70]: kk thanks, I think that makes sense

**Review by tomsilver — APPROVED** [review]
very nice work!

**Pull request merged**
